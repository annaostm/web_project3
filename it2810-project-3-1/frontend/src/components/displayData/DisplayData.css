.pagination {
    display: flex;
    justify-content: center;
    align-items: center;
    margin: 0 auto;
    margin-bottom: 20px;
}

.netflix-cards {
    place-items: center;
    display: grid;
    grid-template-columns: repeat(5, minmax(230px, 1fr));
    grid-gap: 30px;
    margin: 0 auto;
    margin-top: 0px;
    padding-bottom: 40px;
}

.content-text {
    font-size: 18px !important;
}

.wait-text {
    text-align: center;
    font-size: 16px;
    font-weight: 700;
    color: white;
}

.button {
    font-weight: 700 !important;
}

.filter_buttons {
    display: flex;
    justify-content: left;
    margin-left: 10px;
    margin-top: 0px;
    margin-bottom: 30px;
}

.link2 {
    font-weight: 700;
    margin-left: 10px;
    margin-bottom: 10px;
    color: black;
    size: 12px;
}

.link2:hover {
    color: red;
}

@media only screen and (max-width: 1200px) {
    .netflix-cards {
        grid-template-columns: repeat(3, minmax(230px, 1fr));
    }
}

@media only screen and (max-width: 900px) {
    .netflix-cards {
        place-items: center;
        grid-template-columns: repeat(2, minmax(230px, 1fr));
    }
}

@media only screen and (max-width: 550px) {
    .netflix-cards {
        place-items: center;
        grid-template-columns: repeat(1, minmax(230px, 1fr));
    }
    .MuiToolbar-root {
        flex-direction: column;
    }
}

