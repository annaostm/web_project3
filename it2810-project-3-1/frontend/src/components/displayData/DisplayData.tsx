import { useQuery, gql } from "@apollo/client";
import { useEffect, useState } from "react";
import TableFooter from "@mui/material/TableFooter";
import TablePagination from "@mui/material/TablePagination";
import TableRow from "@mui/material/TableRow";
import Card from "@mui/material/Card";
import CardActions from "@mui/material/CardActions";
import CardContent from "@mui/material/CardContent";
import CardMedia from "@mui/material/CardMedia";
import Typography from "@mui/material/Typography";
import movie from "../../images/movie.jpeg";
import s_letter from "../../images/s_letter.png";
import TablePaginationActions from "../pagination/Pagination";
import { useReactiveVar } from "@apollo/client";
import { sortStorageVar, filterStorageVar } from "../filterButton/FilterButton";
import { Link } from "react-router-dom";
import { GET_DATA } from "../../graphql/query";
import "./DisplayData.css";


//Movie interface that defines the type of the netflix object
export interface Data {
  show_id: string;
  title: string;
  type: string;
  release_year: number;
  rating: string;
  country: string;
  duration: string;
  listed_in: string[];
  image_url: string;
}

//NetFlix type that defines the type of the netflix object
type NetflixList = {
  netflix: {
    count: number; //number of total movie/serie objects
    data: Data[]; //array of movie/serie objects
  };
};

//dataProps interface for the DisplayData component
interface dataProps {
  type?: string;
  input?: string;
}

//DisplayData is the component that displays all the movies and series in a table
export default function DisplayData(props: dataProps) {
  //Setting the limit and offset values for the pagination
  const [batchSize, setBatchSize] = useState<number>(10);
  const [pageNumber, setPageNumber] = useState<number>(0);

  //Getting global state variables from the filter buttons
  const sortStorage = useReactiveVar(sortStorageVar);
  const filterStorage = useReactiveVar(filterStorageVar);

  //Reset sortStorage and filterStorage when the type changes
  useEffect(() => {
    sortStorageVar("");
    filterStorageVar("");
  }, [props.type]);

  //Updating the pagination when serach input changes or when the filter or sort buttons are clicked
  useEffect(() => {
    setPageNumber(0);
  }, [props.input, sortStorage, filterStorage]);

  //useQuery hook that gets the data from the database
  const { loading, error, data } = useQuery<NetflixList>(GET_DATA, {
    variables: {
      limit: batchSize,
      offset: pageNumber * batchSize,
      type: props.type,
      title: props.input,
      sort: sortStorage,
      category: filterStorage,
    },
  });

  //Error handling
  if (loading) return <p className="wait-text">Loading...</p>; //If loading
  if (error) return <p className="wait-text">Error :(</p>; //If error
  if (data && data?.netflix.count < 1)
    return <p className="wait-text">No content matched your search</p>; //If no data

  //Function that handles the page change
  const handleChangePage = (
    event: React.MouseEvent<HTMLButtonElement> | null,
    newPage: number
  ) => {
    setPageNumber(newPage);
  };

  //Function that handles the batch size change
  const handleChangeRowsPerPage = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    setBatchSize(parseInt(event.target.value, 10));
    setPageNumber(0);
  };


  let no_image = props.type === "Movie" ? movie : s_letter

  return (
    <div>
      <div className="netflix-cards" aria-label="Nextflix cards">
        {data?.netflix.data.map((netflix: Data) => (
          <div key={netflix.show_id}>
            <div aria-label="A Movie/A Serie">
              <Card sx={{ width: 200 }}>
                <CardMedia
                  component="img"
                  height="270"
                  image={netflix.image_url ? "https://www.themoviedb.org/t/p/w300/" + netflix.image_url : no_image}
                  alt="movie/serie image"
                />
                <CardContent className="card-content" sx={{ height: 140 }}>
                  <Typography
                    className="content-text"
                    gutterBottom
                    variant="h5"
                    component="div"
                  >
                    {netflix.title}
                  </Typography>
                  <Typography variant="body2" color="text.secondary">
                    {"Released: " + netflix.release_year} <br></br>{" "}
                    {"Categories: " + netflix.listed_in}
                  </Typography>
                </CardContent>
                <CardActions>
                  <Link onClick={() => console.log(netflix.show_id)} to={netflix.show_id} className="link2" data-cy="about-link">
                    Learn More
                  </Link>
                </CardActions>
              </Card>
            </div>
          </div>
        ))}
      </div>
      <div className="pagination">
        <TableFooter>
          <TableRow>
            <TablePagination
              rowsPerPageOptions={[
                5,
                10,
                25,
                { label: "All", value: data?.netflix.count || 0 },
              ]}
              colSpan={3}
              count={data?.netflix.count ?? 0}
              rowsPerPage={batchSize}
              page={pageNumber}
              SelectProps={{
                inputProps: {
                  "aria-label": "rows per page",
                },
                native: true,
              }}
              sx={{
                color: "white",
                backgroundColor: "#191b20",
                "& .MuiTablePagination-selectIcon	": {
                  color: "white",
                },
                borderBottom: 0,
              }}
              onPageChange={handleChangePage}
              onRowsPerPageChange={handleChangeRowsPerPage}
              ActionsComponent={TablePaginationActions}
            />
          </TableRow>
        </TableFooter>
      </div>
    </div>
  );
}
