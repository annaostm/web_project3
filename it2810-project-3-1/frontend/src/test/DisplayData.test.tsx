import { gql } from "@apollo/client";
import { unmountComponentAtNode } from "react-dom";
import renderer from "react-test-renderer";
import DisplayData from "../components/displayData/DisplayData";
import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import { GET_DATA } from "../graphql/query";
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect'
import '@testing-library/jest-dom'


// mocks for the apollo client
const mocks = [
  {
    request: {
      query: GET_DATA,
      variables: {
        limit: 10,
        offset: 0,
        type: "Movie",
        title: "Jaws",
        sort: "title",
        category: "",
      }

    },

    result: {
      data: {
        netflix: [
        {"show_id":"s42","title":"Jaws","type":"Movie","release_year":"1975","rating":"PG", "duration":"124 min", "listed_in":["Action & Adventure", "Classic Movies", "Dramas"]},
        {"show_id":"s43","title":"Jaws 2","type":"Movie","release_year":"1978","rating":"PG", "duration":"116 min", "listed_in":["Dramas", " Horror Movies", " Thrillers"]},
        {"show_id":"s44","title":"Jaws 3","type":"Movie","release_year":"1983","rating":"PG", "duration":"98 min", "listed_in":["Action & Adventure", " Horror Movies", " Thrillers"]},
        {"show_id":"s45","title":"Jaws: The Revenge","type":"Movie","release_year":"1987","rating":"PG-13", "duration":"91 min", "listed_in":["Action & Adventure", " Horror Movies", " Thrillers"]},
      ]
      }
    }
  }
];


it("renders without error", async () => {
  render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <DisplayData type={"Movie"} input={"Jaws"} />
    </MockedProvider>
  );
  expect(await screen.findByText('Loading...')).toBeInTheDocument();
  
});

// mocks for the apollo client
const mocks2 = [
  {
    request: {
      query: gql`
        query Query {
          netflix {
            show_id
            title
            type
            director
          }
        }
      `,
    },
    result: {
      data: {
        netflix: [
          {
            show_id: 1,
            title: "Jaws",
            type: "Movie",
            director: "director",
          },
        ],
      },
    },
  },
];

// setting up the mock for the apollo client

let container: Element | DocumentFragment | null = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

// cleanup on exiting
afterEach(() => {
  if (container == null) {
    throw "container is null";
  }
  unmountComponentAtNode(container);
  container = null;
});

// Test for DisplayData
it("Test for DisplayData", () => {
  const tree = renderer
    .create(
      <MockedProvider mocks={mocks2}>
        <DisplayData type={"Movie"} input={"Jaws"} />
      </MockedProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
