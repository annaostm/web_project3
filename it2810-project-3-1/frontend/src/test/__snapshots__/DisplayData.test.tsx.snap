// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for DisplayData 1`] = `
<p
  className="wait-text"
>
  Loading...
</p>
`;
